libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Memory/Scopes/BotStateMemoryScope.cs

```diff
@@ -46,7 +40,8 @@ public override object GetMemory(DialogContext dialogContext)
                 throw new ArgumentNullException($"{nameof(dialogContext)} is null");
             }
 
-            return this.state;
+            var cachedState = dialogContext.Context.TurnState.Get<object>(typeof(T).Name);
+            return cachedState.GetType().GetProperty("State").GetValue(cachedState);
```

Comment:
Not a fan of this but I understand why we need it. If I come up with an alternative I'll ping you.